### DISCUSSION TITLE
How to store and use the date and time

### DISCUSSION THREAD

**Discussion by Michael-pi-projects** [Makuna/Rtc#140] [Q&A] [answered]
Hi !!
I am new here .First thanks for this Rtc library and etc , I am using it ( more specific the with the DS3231 module and with the example DS3231 simple ) in a project of mine and i want to know if is there any way to get the time and date and store it to a variable or how else to store  and use it for other perpose , for example to store it on an sd every 3 seconds.
I am new to arduino , I only have used it and code about 3 to 4 months with basic and simple projects and basic senros.
thanks!!!

**Comment by Makuna** [comment, marked as answer]
The current date time returned from the RTC is a RtcDateTime object.

The easiest way to store this is to call `Epoch32Time()` which returns a 32bit value that you can store.
Then you retrieve that stored value and pass to `InitWithEpoch32Time(uint32_t time)` to create an RtcDateTime object again.

Epoch time is a standard, the RtcDateTime object supports both 32 bit and 64 bit versions.

**Reply by Michael-pi-projects** [reply]
thanks o lot !! But is it posible to give an example of the code how to do it because i am not to familiar with objects on arduino ;
Or if you want i can show mine  code  here to get a better understanding .

**Reply by Michael-pi-projects** [reply]
//Here is the code and I want in the void loop to store the date and time
//Icomented the temperature because I dont want to use it


#include <Wire.h> // must be included here so that Arduino library object file references work
#include <RtcDS3231.h>

RtcDS3231<TwoWire> Rtc(Wire);

void DATE_TIME_YEAR(){
  {
    if (!Rtc.IsDateTimeValid()) 
    {
        if (Rtc.LastError() != 0)
        {
            // we have a communications error
            // see https://www.arduino.cc/en/Reference/WireEndTransmission for 
            // what the number means
            Serial.print("RTC communications error = ");
            Serial.println(Rtc.LastError());
        }
        else
        {
            // Common Causes:
            //    1) the battery on the device is low or even missing and the power line was disconnected
            Serial.println("RTC lost confidence in the DateTime!");
        }
    }

    RtcDateTime now = Rtc.GetDateTime();
    printDateTime(now);
    Serial.println();   


	/*RtcTemperature temp = Rtc.GetTemperature();
	temp.Print(Serial);
	// you may also get the temperature as a float and print it
  // Serial.print(temp.AsFloatDegC());
  Serial.println("C");

  delay(10000); ten seconds*/
  //return(now);
  }
}
#define countof(a) (sizeof(a) / sizeof(a[0]))

void printDateTime(const RtcDateTime& dt){
    char datestring[20];

    snprintf_P(datestring, 
            countof(datestring),
            PSTR("%02u/%02u/%04u,%02u:%02u:%02u"),
            dt.Month(),
            dt.Day(),
            dt.Year(),
            dt.Hour(),
            dt.Minute(),
            dt.Second() );
    Serial.print(datestring);Serial.print(" ");
}

void setup() {
 Serial.begin(115200);
 Serial.print("compiled: ");
    Serial.print(__DATE__);
    Serial.println(__TIME__);

    //--------RTC SETUP ------------
    // if you are using ESP-01 then uncomment the line below to reset the pins to
    // the available pins for SDA, SCL
    // Wire.begin(0, 2); // due to limited pins, use pin 0 and 2 for SDA, SCL
    
    Rtc.Begin();

    RtcDateTime compiled = RtcDateTime(__DATE__, __TIME__);
    printDateTime(compiled);
    Serial.println();

    if (!Rtc.IsDateTimeValid()) 
    {
        if (Rtc.LastError() != 0)
        {
            // we have a communications error
            // see https://www.arduino.cc/en/Reference/WireEndTransmission for 
            // what the number means
            Serial.print("RTC communications error = ");
            Serial.println(Rtc.LastError());
        }
        else
        {
            // Common Causes:
            //    1) first time you ran and the device wasn't running yet
            //    2) the battery on the device is low or even missing

            Serial.println("RTC lost confidence in the DateTime!");

            // following line sets the RTC to the date & time this sketch was compiled
            // it will also reset the valid flag internally unless the Rtc device is
            // having an issue

            Rtc.SetDateTime(compiled);
        }
    }

    if (!Rtc.GetIsRunning())
    {
        Serial.println("RTC was not actively running, starting now");
        Rtc.SetIsRunning(true);
    }

    RtcDateTime now = Rtc.GetDateTime();
    if (now < compiled) 
    {
        Serial.println("RTC is older than compile time!  (Updating DateTime)");
        Rtc.SetDateTime(compiled);
    }
    else if (now > compiled) 
    {
        Serial.println("RTC is newer than compile time. (this is expected)");
    }
    else if (now == compiled) 
    {
        Serial.println("RTC is the same as compile time! (not expected but all is fine)");
    }

    // never assume the Rtc was last configured by you, so
    // just clear them to your needed state
    Rtc.Enable32kHzPin(false);
    Rtc.SetSquareWavePin(DS3231SquareWavePin_ModeNone); 
}

void loop() {
   DATE_TIME_YEAR();
   delay(3000);
}

**Reply by Makuna** [reply]
You need to work out how to store and retrieve the uint32_t, but here is the snippits of code to work with the RtcDateTime

to get a uint32_t from RTC module...
```
RtcDateTime now = Rtc.GetDateTime();  // reteives the current time from the module
uint32_t epochTime = now.Epoch32Time(); // converts the time stored in 'now' to uint32_t epoch time standard value of 32 bits
// now just store epochTime however you want
```

to get a RtcDateTime from uint32_t is just...
```
uint32_t epochTime = // however your set or retrieve it
RtcDateTime time;

time.InitWithEpoch32Time( epochTime );
// now do what ever you want with time.
```

**Reply by Michael-pi-projects** [reply]
you meen i ca use time variable as i want ?

one more question : what type of data does the time return ? and what type of data does the epochTime return ?

**Reply by Makuna** [reply]
search internet for the specific term "epoch time",  it's a standard, its seconds since a specific date and time.
